### ISSUE TITLE
Decide what to do about `UnsafePinned` and safe `Pin::as_ref`

### ISSUE THREAD

**Issue by traviscross** [rust-lang/rust#137750]
As context, this entirely safe code is reported as UB by Miri under both Stacked Borrows and Tree Borrows:

```rust
use core::{ops::Deref, pin::{pin, Pin}, task::{Context, Poll, Waker}};

fn main() {
    let mut f = pin!(async move {
        let x = &mut 0u8;
        core::future::poll_fn(move |_| {
            *x = 1;
            //~^ ERROR write access is forbidden
            Poll::<()>::Pending
        }).await
    });
    let mut cx = Context::from_waker(&Waker::noop());
    assert_eq!(f.as_mut().poll(&mut cx), Poll::Pending);
    _ = <Pin<&mut _> as Deref>::deref(&f);
    assert_eq!(f.as_mut().poll(&mut cx), Poll::Pending);
}
```

[Playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&code=%2F%2F+This+demonstrates+something+that+Miri+reports+as+UB+under+both%0A%2F%2F+Stacked+Borrows+and+Tree+Borrows+even+though+it+consists+entirely%0A%2F%2F+of+safe+code.%0A%2F%2F%0A%2F%2F+This+is+due+to+a+fundamental+unsoundness+in+our+model+that+we%27re%0A%2F%2F+trying+to+fix+with+%60UnsafePinned%60%2C+but+this+example+shows+that+even%0A%2F%2F+that%2C+as+RFCed%2C+may+not+quite+be+enough+because+of+how+1%29+creating%0A%2F%2F+a+shared+reference+acts+as+a+%22read%22+of+all+data+accessible+behind%0A%2F%2F+that+shared+reference+%28except+for+%60UnsafeCell%3C_%3E%60+types%29+and+2%29+we%0A%2F%2F+stabilized+%60%3CPin+as+Deref%3E%3A%3Aderef%60+as+a+safe+function.%0A%2F%2F%0A%2F%2F+Author%3A+TC%0A%2F%2F+Date%3A+2025-02-27%0A%2F%2F%0A%2F%2F+See%3A%0A%2F%2F%0A%2F%2F+-+https%3A%2F%2Fgithub.com%2Frust-lang%2Frust%2Fissues%2F137750%0A%2F%2F%0A%2F%2F+Run+with%3A%0A%2F%2F%0A%2F%2F+++++MIRIFLAGS%3D%22-Zmiri-tree-borrows%22+cargo+miri+run%0A%2F%2F+++++cargo+miri+run%0A%0Ause+core%3A%3A%7Bops%3A%3ADeref%2C+pin%3A%3A%7Bpin%2C+Pin%7D%2C+task%3A%3A%7BContext%2C+Poll%2C+Waker%7D%7D%3B%0A%0Afn+main%28%29+%7B%0A++++let+mut+f+%3D+pin%21%28async+move+%7B%0A++++++++let+x+%3D+%26mut+0u8%3B%0A++++++++core%3A%3Afuture%3A%3Apoll_fn%28move+%7C_%7C+%7B%0A++++++++++++%2Ax+%3D+1%3B%0A++++++++++++%2F%2F~%5E+ERROR+write+access+is+forbidden%0A++++++++++++Poll%3A%3A%3C%28%29%3E%3A%3APending%0A++++++++%7D%29.await%0A++++%7D%29%3B%0A++++let+mut+cx+%3D+Context%3A%3Afrom_waker%28%26Waker%3A%3Anoop%28%29%29%3B%0A++++assert_eq%21%28f.as_mut%28%29.poll%28%26mut+cx%29%2C+Poll%3A%3APending%29%3B%0A++++_+%3D+%3CPin%3C%26mut+_%3E+as+Deref%3E%3A%3Aderef%28%26f%29%3B%0A++++assert_eq%21%28f.as_mut%28%29.poll%28%26mut+cx%29%2C+Poll%3A%3APending%29%3B%0A%7D%0A)

<details>
<summary>Output</summary>
Under TB:

```rust
error: Undefined Behavior: write access through <1666> at alloc969[0x8] is forbidden
  --> src/main.rs:7:13
   |
7  |             *x = 1;
   |             ^^^^^^ write access through <1666> at alloc969[0x8] is forbidden
   |
   = help: this indicates a potential bug in the program: it performed an invalid operation, but the Tree Borrows rules it violated are still experimental
   = help: the accessed tag <1666> has state Frozen which forbids this child write access
help: the accessed tag <1666> was created here, in the initial state Reserved
  --> src/main.rs:6:9
   |
6  | /         core::future::poll_fn(move |_| {
7  | |             *x = 1;
8  | |             Poll::<()>::Pending
9  | |         }).await
   | |________________^
help: the accessed tag <1666> later transitioned to Active due to a child write access at offsets [0x8..0x9]
  --> src/main.rs:7:13
   |
7  |             *x = 1;
   |             ^^^^^^
   = help: this transition corresponds to the first write to a 2-phase borrowed mutable reference
help: the accessed tag <1666> later transitioned to Frozen due to a reborrow (acting as a foreign read access) at offsets [0x0..0x10]
  --> src/main.rs:13:9
   |
13 |     _ = <Pin<&mut _> as Deref>::deref(&f);
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   = help: this transition corresponds to a loss of write permissions
   = note: BACKTRACE (of the first span):
   = note: inside closure at src/main.rs:7:13: 7:19
  --> src/main.rs:9:12
   |
9  |         }).await
   |            ^^^^^
note: inside `main`
  --> src/main.rs:14:16
   |
14 |     assert_eq!(f.as_mut().poll(&mut cx), Poll::Pending);
   |                ^^^^^^^^^^^^^^^^^^^^^^^^

note: some details are omitted, run with `MIRIFLAGS=-Zmiri-backtrace=full` for a verbose backtrace

error: aborting due to 1 previous error
```
</details>

This isn't a bug in Miri really.  It's a bug... elsewhere.  Part of that is described by:

- https://github.com/rust-lang/rust/issues/63818

But at the same time, this is relevant to the post-RFC work on `UnsafePinned`:

- https://github.com/rust-lang/rust/issues/125735

The idea of `UnsafePinned` is that we use it to wrap the fields subject to self-borrows so that given a mutable reference to such fields, the compiler (and our model) won't assume that a foreign write can't occur.

But the RFC suggested that while `&mut UnsafePinned<T>` was special, `&UnsafePinned<T>` wasn't, and would therefore act like `&T`.  This is where we've hit problems due to safe `Pin::deref`, as @RalfJung has described:

- https://github.com/rust-lang/rust/issues/125735#issuecomment-2299183374
- https://github.com/rust-lang/rust/pull/137043#discussion_r1973978597
- https://github.com/rust-lang/miri/issues/3796#issuecomment-2299177277

In the above example, if the compiler were to insert:

```rust
let x: &mut UnsafePinned<0u8> = ..;
```

Then we still end up having a problem, under the RFC's model and our current model behavior, which is (quoting @RalfJung):

> when a shared reference gets created, do a read of everything the reference points to (except `UnsafeCell`) to ensure things are in a sane state

What happens in the example is that:

- We create and store a mutable self-reference to the `u8` owned by the future.
- We activate that self-reference with a write, marking it `Unique` under TB.
- Then, due to the `deref` creating a reference to the future and our behavior to treat that as a read of the entire future, a foreign read occurs, marking the mutable reference as `Frozen`.
- Next time we try to write, we get UB.

About this, @RalfJung has [said](https://github.com/rust-lang/rust/issues/125735#issuecomment-2299183374):

> In the RFC we had the question whether `UnsafePinned` should also act like an `UnsafeCell`. I am now fairly sure that the answer is "yes, it must". The reason is [this](https://github.com/rust-lang/miri/issues/3796#issuecomment-2299177277): due to `Pin::deref` being safe, it is at any moment possible to create a shared reference to a future that is halted at a suspension point. If there is no `UnsafeCell`, this acts like a read of the entire future data, and that read conflicts with mutable references that the future may hold to its own data. This is not a necessary property of aliasing with mutable reference, but it is a necessary consequence of the decision that `Pin::deref` should be safe.

And also [that](https://github.com/rust-lang/rust/pull/137043#discussion_r1973978597):

> One could imagine aliasing models that don't do such a read, though then `dereferenceable` becomes more questionable and I don't know how this would impact optimizations.

Anyway, it seems we should have a canonical issue to use to decide about how to proceed about this and how either our model or RFC 3467 would need to be adjusted before the stabilization of `UnsafePinned`, so here is that issue.

---

#### 2025-04-26:

As a slight refinement here, note that we actually don't even need `deref`, as `Pin::as_ref` is enough.

```rust
use core::{
    pin::{pin, Pin},
    task::{Context, Poll, Waker},
};

fn main() {
    let mut f = pin!(async move {
        let x = &mut 0u8;
        core::future::poll_fn(move |_| {
            *x = 1;
            //~^ ERROR write access is forbidden
            Poll::<()>::Pending
        })
        .await
    });
    let mut cx = Context::from_waker(&Waker::noop());
    assert_eq!(f.as_mut().poll(&mut cx), Poll::Pending);
    let _: Pin<&_> = f.as_ref(); // Or: `f.as_mut().into_ref()`.
    assert_eq!(f.as_mut().poll(&mut cx), Poll::Pending);
}
```

[Playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&code=%2F%2F+This+demonstrates+something+that+Miri+reports+as+UB+under+both%0A%2F%2F+Stacked+Borrows+and+Tree+Borrows+even+though+it+consists+entirely%0A%2F%2F+of+safe+code.%0A%2F%2F%0A%2F%2F+This+is+due+to+a+fundamental+unsoundness+in+our+model+that+we%27re%0A%2F%2F+trying+to+fix+with+%60UnsafePinned%60%2C+but+this+example+shows+that+even%0A%2F%2F+that%2C+as+RFCed%2C+may+not+quite+be+enough+because+of+how+1%29+creating%0A%2F%2F+a+shared+reference+acts+as+a+%22read%22+of+all+data+accessible+behind%0A%2F%2F+that+shared+reference+%28except+for+%60UnsafeCell%3C_%3E%60+types%29+and+2%29+we%0A%2F%2F+stabilized+%60Pin%3A%3Aas_ref%60+and+%60Pin%3A%3Ainto_ref%60+as+a+safe+functions.%0A%2F%2F%0A%2F%2F+Author%3A+TC%0A%2F%2F+Date%3A+2025-04-26%0A%2F%2F%0A%2F%2F+See%3A%0A%2F%2F%0A%2F%2F+-+https%3A%2F%2Fgithub.com%2Frust-lang%2Frust%2Fissues%2F137750%0A%2F%2F%0A%2F%2F+Run+with%3A%0A%2F%2F%0A%2F%2F+++++MIRIFLAGS%3D%22-Zmiri-tree-borrows%22+cargo+miri+run%0A%2F%2F+++++cargo+miri+run%0A%0Ause+core%3A%3A%7B%0A++++pin%3A%3A%7Bpin%2C+Pin%7D%2C%0A++++task%3A%3A%7BContext%2C+Poll%2C+Waker%7D%2C%0A%7D%3B%0A%0Afn+main%28%29+%7B%0A++++let+mut+f+%3D+pin%21%28async+move+%7B%0A++++++++let+x+%3D+%26mut+0u8%3B%0A++++++++core%3A%3Afuture%3A%3Apoll_fn%28move+%7C_%7C+%7B%0A++++++++++++%2Ax+%3D+1%3B%0A++++++++++++%2F%2F~%5E+ERROR+write+access+is+forbidden%0A++++++++++++Poll%3A%3A%3C%28%29%3E%3A%3APending%0A++++++++%7D%29%0A++++++++.await%0A++++%7D%29%3B%0A++++let+mut+cx+%3D+Context%3A%3Afrom_waker%28%26Waker%3A%3Anoop%28%29%29%3B%0A++++assert_eq%21%28f.as_mut%28%29.poll%28%26mut+cx%29%2C+Poll%3A%3APending%29%3B%0A++++let+_%3A+Pin%3C%26_%3E+%3D+f.as_ref%28%29%3B+%2F%2F+Or%3A+%60f.as_mut%28%29.into_ref%28%29%60.%0A++++assert_eq%21%28f.as_mut%28%29.poll%28%26mut+cx%29%2C+Poll%3A%3APending%29%3B%0A%7D%0A)

---

cc @RalfJung @WaffleLapkin @rust-lang/opsem @rust-lang/lang

@rustbot labels +T-lang +T-opsem +C-discussion +F-unsafe_pinned

**Comment by traviscross** [comment]
Regarding this bit:

> One could imagine aliasing models that don't do such a read, though then `dereferenceable` becomes more questionable and I don't know how this would impact optimizations.

@RalfJung: Does it work at all, and if so, how ugly would it be to otherwise treat `&UnsafePinned<T>` like `&T` in the model but to not do this read in this one particular case?

**Comment by CAD97** [comment]
It would defer the unsoundness, but safe code would still be allowed to do real source reads through the `&{async block}` (e.g. by casting to `&[MaybeUninit<u8>]`). The borrow relaxations required are *almost exactly* the same as with `UnsafeCell`, as we have aliasing `&Outer` and `&mut Inner` simultaneously.

The difference is that the `&mut` in the async block is "inactive" when the `&Outer` exists and gets reactivated by passing `&mut Outer` to `Future::poll`… but this difference is only skin deep. Pinning that exposes `&mut`-derived access outside of `Future::poll` is *difficult* to write soundly, but it should still be considered valid, as long as dropping the containing future still terminates the access. Scoped async tasks would be a canonical instance of such[^1], if they could be restricted to borrowing from the parent task.

I believe that theoretically it would be coherent to add a `TempFrozen` retag that is invalidated by sibling `Unique` (unless unprotected) instead of downgrading sibling `Unique` when created, but the model and reasoning complexity costs that requires don't seem all that beneficial. (The only potential "benefit" I can really see is adding new ways in which scoped async spawn and borrowing completion based async IO are both fundamentally unsound in the face of careless but reasonable seeming safe code[^1].)

[^1]: Scoped async tasks are believed to be sound if all borrowed memory is owned by the parent task and drop-cancelling the parent task waits on drop-cancelling the child tasks. With `TempFrozen` semantics, however, this is not true.

**Comment by RalfJung** [comment]
> Does it work at all, and if so, how ugly would it be to otherwise treat &UnsafePinned<T> like &T in the model but to not do this read in this one particular case? (So that, e.g., if it affects optimizations, it affects fewer places.)

In SB, this is just plain impossible.

In TB, we could treat such a reference like `&()`, which reads nothing but as TB allows accessing pointers "outside" the range of memory their type indicates, could still be used to access the underlying future. We'd have to remove the `dereferenceable` attribute on such a reference to match that.

But... honestly it doesn't seem worth it. Allowing aliasing of `&mut` in particular for `Pin<&mut T>` is the entire point of `UnsafePinned`. We decided a long time ago that `Pin<&mut T>` gives safe access to `&T`.  We should just accept the consequence that this means `&T` is part of subtle aliasing games, and stop having `noalias` on that memory.
